### PR TITLE
Fix bug which can't delete legacy rules and groups

### PR DESCRIPTION
**Pull request by zhengxiexie**
Delete the legacy rule and group at last when updating security policy. 

Use InfraClient to refactor CRUD.
Split large file.
Based on v4.0.0 branch.

### CODE REVIEW

File: pkg/nsx/services/store.go [hunk 1]
```diff
@@ -95,19 +95,15 @@ func queryGroup(service *SecurityPolicyService, wg *sync.WaitGroup, fatalErrors
 		if err != nil {
 			fatalErrors <- err
 		}
-		typeConverter := service.NSXClient.RestConnector.TypeConverter()
 		for _, g := range response.Results {
-			a, err := typeConverter.ConvertToGolang(g, model.GroupBindingType())
+			a, err := Converter.ConvertToGolang(g, model.GroupBindingType())
```
Comment (heypnus) [inline pkg/nsx/services/store.go:99]: Where is the defination of `Converter`?

File: pkg/nsx/services/store.go [hunk 2]
```diff
@@ -95,19 +95,15 @@ func queryGroup(service *SecurityPolicyService, wg *sync.WaitGroup, fatalErrors
...
 			if err != nil {
 				for _, e := range err {
 					fatalErrors <- e
 				}
 			}
 			c, _ := a.(model.Group)
-			err2 := service.GroupStore.Add(c)
-			if err2 != nil {
-				fatalErrors <- err2
-			}
```
Comment (heypnus) [inline pkg/nsx/services/store.go:110]: Why remove the error detect for `service.GroupStore.Add`?

File: pkg/nsx/services/store.go [hunk 3]
```diff
@@ -230,45 +216,93 @@ func transError(err error) error {
...
+			return err
+		}
 	}
-	var existingRules []model.Rule
-	for _, rule := range indexResults {
-		existingRules = append(existingRules, rule.(model.Rule))
+	return nil
+}
+
+func (service *SecurityPolicyService) OperateRulesStore(sp *model.SecurityPolicy) error {
```
Comment (heypnus) [inline pkg/nsx/services/store.go:239]: `OperateRulesStore` -> `OperateRuleStore`

File: pkg/nsx/services/store.go [hunk 4]
```diff
@@ -230,45 +216,93 @@ func transError(err error) error {
...
-	return existingGroups, &existingSecurityPolicy, existingRules, nil
+	return nil
 }
 
-func (service *SecurityPolicyService) ListSecurityPolicyID() sets.String {
-	groups := service.GroupStore.ListIndexFuncValues(util.TagScopeSecurityPolicyCRUID)
-	groupSet := sets.NewString()
-	for _, group := range groups {
-		groupSet.Insert(group)
+func (service *SecurityPolicyService) OperateGroupsStore(gs *[]model.Group) error {
```
Comment (heypnus) [inline pkg/nsx/services/store.go:258]: ditto.
And why in `OperateSecurityStore` and `OperateRulesStore` the parameter is `sp *model.SecurityPolicy` but `OperateGroupsStore` didn't follow that?

File: pkg/nsx/services/wrap.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+	"github.com/vmware/vsphere-automation-sdk-go/services/nsxt/model"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/util"
+)
+
+// Patch API at infra level can be used in two flavours.
+// 1. Like a regular API to update Infra object.
+// 2. Hierarchical API: To create/update/delete entire or part of intent hierarchy Hierarchical.
+// We use infra patch API in hierarchical mode to create/update/delete entire or part of intent hierarchy,
+// for this convenience we can no longer // CRUD CR separately, and reduce the number of API calls to NSX-T.
```
Comment (heypnus) [inline pkg/nsx/services/wrap.go:14]: What's the meaning of `//` in the middle of this line?

File: pkg/controllers/securitypolicy_controller.go [hunk 6]
```diff
@@ -260,33 +273,45 @@ func (r *SecurityPolicyReconciler) GarbageCollector(cancel chan bool, timeout ti
 			return
 		case <-time.After(timeout):
 		}
-		nsxPolicySet := r.Service.ListSecurityPolicyID()
-		if len(nsxPolicySet) == 0 {
+
+		keys := r.Service.ListSecurityPolicyKeys()
```
Comment (dantingl) [inline pkg/controllers/securitypolicy_controller.go:277]: Could you help explain why you update logic in GarbageCollector()? I did not find anything different or better between existing and new implementation.

File: pkg/util/utils.go [hunk 7]
```diff
@@ -59,6 +59,15 @@ func RemoveDuplicateStr(strSlice []string) []string {
 	return resultStr
 }
 
+func Contains(s []string, str string) bool {
```
Comment (heypnus) [inline pkg/util/utils.go:62]: Where is the func `Contains` used?

Comment (dantingl) [inline pkg/util/utils.go:62]: Looks like `containsString` in securitypolicy_controller.go is not needed?

File: pkg/config/config.go [hunk 8]
```diff
@@ -158,7 +158,7 @@ func (operatorConfig *NSXOperatorConfig) GetTokenProvider() auth.TokenProvider {
 }
 
 func (operatorConfig *NSXOperatorConfig) createTokenProvider() error {
-	log.V(2).Info("try to load VC host CA")
+	log.V(1).Info("try to load VC host CA")
```
Comment (dantingl) [inline pkg/config/config.go:161]: @TaoZou1 do we need to change log level here?

File: docs/security-policy.md [hunk 9]
```diff
@@ -284,7 +284,8 @@ In the same policy, the higher rule has the higher priority. E.g. in the policy:
         - podSelector: {}
 ...
 ```
-There're 3 rules in the array, rule[0] allows connections from Pods with the label
+
```
Comment (dantingl) [inline docs/security-policy.md:287]: I prefer not to include unnecessary changes

File: pkg/controllers/securitypolicy_controller.go [hunk 10]
```diff
@@ -121,7 +121,7 @@ func (r *SecurityPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 	} else {
 		if controllerutil.ContainsFinalizer(obj, util.FinalizerName) {
 			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, METRIC_RES_TYPE)
-			if err := r.Service.DeleteSecurityPolicy(obj.UID); err != nil {
+			if err := r.Service.DeleteSecurityPolicy(types.UID(obj.UID)); err != nil {
```
Comment (heypnus) [inline pkg/controllers/securitypolicy_controller.go:124]: Just curious that why it's `(types.UID(obj.UID))`, not `(obj.UID)`?

File: pkg/nsx/services/compare.go [hunk 11]
```diff
@@ -2,93 +2,109 @@ package services
...
-	// sort the rules by id, otherwise expandRule may return different results, only the sequence of the
-	// rule is different, so sort by port number, and it avoids the needless updates.
-	var sortRules = func(rules []model.Rule) {
-		sort.Slice(rules, func(i, j int) bool {
-			return *(rules[i].Id) > *(rules[j].Id)
-		})
-	}
-	sortRules(existingRules)
-	sortRules(rules)
+func (service *SecurityPolicyService) rulesEqual(existingRules []model.Rule, rules []model.Rule) ([]model.Rule, []model.Rule) {
```
Comment (heypnus) [inline pkg/nsx/services/compare.go:21]: Can we rename `rulesEqual` and `groupsEqual` to `rulesCompare` and `groupsCompare`? Because I feel that the function xxxEqual should return a value of boolean type...

File: pkg/controllers/securitypolicy_controller.go (hunk 10 above)
Comment (zhengxiexie) [reply pkg/controllers/securitypolicy_controller.go:124]: Yes, it's redundant, changed.

File: pkg/nsx/services/compare.go (hunk 11 above)
Comment (zhengxiexie) [reply pkg/nsx/services/compare.go:21]: ok

File: pkg/nsx/services/firewall.go [hunk 12]
```diff
@@ -35,192 +34,183 @@ type SecurityPolicyService struct {
...
+			log.Error(err, "failed to build SecurityPolicy")
+			return err
+		}
+	case types.UID:
+		// We can delete the security policy directly by its ID,
+		// It's related resources will be deleted automatically,
+		// So don't worry we have no nsxGroups and nsxRules.
+		indexResults, err := service.SecurityPolicyStore.ByIndex(util.TagScopeSecurityPolicyCRUID, string(sp))
+		if err != nil {
+			log.Error(err, "failed to get security policy", "UID", string(sp))
```
[the remark this answers is not in the document]
Comment (zhengxiexie) [reply pkg/nsx/services/firewall.go:161]: Talked as in slack.

**Review by heypnus — APPROVED** [review]
